File: libraries/common_network/src/main/java/com/tokopedia/common/network/data/source/cache/RestCacheDataStore.java
package com.tokopedia.common.network.data.source.cache;

import com.tokopedia.common.network.data.model.RestRequest;
import com.tokopedia.common.network.data.model.RestResponseIntermediate;
import com.tokopedia.common.network.data.source.RestDataStore;
import com.tokopedia.common.network.util.CommonUtil;
import com.tokopedia.common.network.util.FingerprintManager;
import com.tokopedia.common.network.util.RestCacheManager;
import com.tokopedia.common.network.util.NetworkClient;
import com.tokopedia.common.network.util.RestConstant;

import javax.inject.Inject;

import rx.Observable;

import static com.tokopedia.common.network.util.RestConstant.RES_CODE_CACHE;

/**
 * Retrieve the response from cache only
 */
public class RestCacheDataStore implements RestDataStore {

    private RestCacheManager mCacheManager;
    private FingerprintManager mFingerprintManager;

    @Inject
    public RestCacheDataStore() {
        this.mCacheManager = new RestCacheManager();
        this.mFingerprintManager = NetworkClient.getFingerPrintManager();
    }

    @Override
    public Observable<RestResponseIntermediate> getResponse(RestRequest request) {
        RestResponseIntermediate returnResponse;
        try {
            String rawJson = mCacheManager.get(mFingerprintManager.generateFingerPrint(request.toString(), request.getCacheStrategy().isSessionIncluded()));

            if (rawJson == null || rawJson.isEmpty()) {
                return Observable.just(null);
            }

            returnResponse = new RestResponseIntermediate(CommonUtil.fromJson(rawJson, request.getTypeOfT()), request.getTypeOfT(), true);
            returnResponse.setCode(RES_CODE_CACHE);
            returnResponse.setError(false);
            return Observable.just(returnResponse);
        } catch (Exception e) {
            //For any kind of error body always be null,
            //E.g. JSONException while serializing json to POJO.
            returnResponse = new RestResponseIntermediate(null, request.getTypeOfT(), true);
            returnResponse.setCode(RestConstant.INTERNAL_EXCEPTION);
            returnResponse.setErrorBody("Caught Exception please fix it--> Responsible class : " + e.getClass().toString() + " Detailed Message: " + e.getMessage() + ", Cause by: " + e.getCause());
            returnResponse.setError(true);
        }

        return Observable.just(returnResponse);
    }
}

File: libraries/common_network/src/main/java/com/tokopedia/common/network/coroutines/datasource/RestCacheDataStore.kt
package com.tokopedia.common.network.coroutines.datasource

import com.tokopedia.common.network.data.model.RestRequest
import com.tokopedia.common.network.data.model.RestResponse
import com.tokopedia.common.network.util.*
import com.tokopedia.common.network.util.RestConstant.RES_CODE_CACHE
import kotlinx.coroutines.experimental.Dispatchers
import kotlinx.coroutines.experimental.withContext

/**
 * Retrieve the response from cache only
 */
class RestCacheDataStore constructor(
        private val mCacheManager: RestCacheManager,
        private val mFingerprintManager: FingerprintManager
) : RestDataStore {

    override suspend fun getResponse(request: RestRequest): RestResponse =
            withContext(Dispatchers.IO){

                val rawJson =  mCacheManager.get(mFingerprintManager.generateFingerPrint(request.toString(),
                        request.cacheStrategy.isSessionIncluded))
                try {
                    if (rawJson == null){
                        RestResponse(Unit, RestConstant.INTERNAL_EXCEPTION, true).apply {
                            type = request.typeOfT
                            errorBody = "No Cache found"
                            isError = true
                        }
                    } else {
                        RestResponse(CommonUtil.fromJson(rawJson, request.typeOfT), RES_CODE_CACHE, true).apply {
                            type = request.typeOfT
                            isError = false
                        }
                    }
                } catch (e: Throwable){
                    RestResponse(Unit, RestConstant.INTERNAL_EXCEPTION, true).apply {
                        type = request.typeOfT
                        errorBody = "Caught Exception please fix it--> Responsible class : " + e.javaClass.toString() +
                                " Detailed Message: " + e.message + ", Cause by: " + e.cause
                        isError = true
                    }
                }
    }


}

